PostSchedule: Add AM/PM selector

obenland opened the pull request:
This PR adds an AM/PM selector to the scheduling component of the
editor.

Fixes #10762.

To test: Make sure your site’s time format is set to include AM/PM
(setting is still in wp-admin). Navigate to the editor and open the
calendar next to the Publish button. Time should adjust in the
background.

gwwar commented on client/components/post-schedule/index.jsx:
```diff
@@ -168,7 +169,8 @@ class PostSchedule extends Component {
 		return (
-			<div className="post-schedule" >
+			<div className="post-schedule">
+				{ this.props.site && <QuerySiteSettings siteId={ this.props.site.ID } /> }
```
Shouldn't this be grouped with Clock instead?

obenland replied:
Right, I was thinking about that but once we have [a way to map `time_format` to a moment.js format](https://github.com/Automattic/wp-calypso/pull/10400/files#diff-df1fc8d02d6305e47e8598afe79e9df8), other child components of `PostSchedule` (like `Header`) will need it

gwwar replied:
👍 Let's at least leave a note that Clock is using this. If we looked by at the file by itself it looks like it's querying data for no reason.

retrofox commented on client/state/selectors/get-site-setting.js:
```diff
@@ -0,0 +1,18 @@
+ */
+import { get } from 'lodash';
+
+import { getSiteSettings } from 'state/site-settings/selectors';
```
we should add `internal dependencies` comment block

gwwar commented on client/components/post-schedule/clock.jsx:
```diff
@@ -81,6 +84,18 @@ class PostScheduleClock extends Component {
 		this.props.onChange( date, modifiers );
+	};
+
+	setAmPm( event, amPm ) {
```
Maybe `amPm` can be `amOrPm`

retrofox commented:
There is a weird behavior when you try to define military time hour: In the following animation, I'm typing number `2` again and again.

![post-schedule-am-pm-02](https://cloud.githubusercontent.com/assets/77539/22845310/c55b44a2-efc0-11e6-8478-82875c68b2ad.gif)

The time is automatically updated to `AM/PM` jumping from `02am` to `10pm` which makes because we are allowing type `22`.

I think we should update the limits when the mode is am/pm or not.

gwwar commented on client/components/post-schedule/clock.jsx:
```diff
@@ -148,15 +163,17 @@ class PostScheduleClock extends Component {
 	render() {
-		const { date } = this.props;
+		const { date, timeFormat, translate } = this.props;
+		const hasAmPm = endsWith( timeFormat, 'A' ) || endsWith( timeFormat, 'a' );
```
Maybe `twelveHourTime`?

obenland commented:
There seems to be all kinds of weird behaviors when trying to type in those fields :(

gwwar commented:
Yeah verified the strange behavior too. Can we have the input always store 24 hour time, and then modify the display when 12 hour time? We can do something like remove the max value limit, and mod by 12 or 24, so we can type more easily. 

Ideally I'd just like to see this get replaced by a clock in the future 😄 I've always disliked these controls.

retrofox commented:
> There seems to be all kinds of weird behaviors when trying to type in those fields :(

yup, It is in some way. But the rule is simple: you should't be able to type any value which doesn't match with a valid hours/minutes valule.
In 24-hours-clock you could type from `00` to `23` hours. Minutes from `00` to `59`

gwwar commented:
Part of this weirdness might be from the `max` html5 prop. Maybe we could validate on input blur?

obenland commented:
Yeah, I'm experimenting a bit more. Ideally users should only be able to type in 1-12 when in am/pm mode, I'm pretty sure it's just a matter of taking that input, and converting that into a valid 24h time for moment to then be able to parse and format again

retrofox commented:
> Yeah, I'm experimenting a bit more. Ideally users should only be able to type in 1-12 when in am/pm mode, I'm pretty sure it's just a matter of taking that input, and converting that into a valid 24h time for moment to then be able to parse and format again

We just need to swtich the value from `24` to `12` when the time format is 12-hours-date.

```js
value = value > 23 ? 0 : value;
```
https://github.com/Automattic/wp-calypso/blob/master/client/components/post-schedule/clock.jsx#L48

retrofox commented:
Kostantin, I've been taking a look to your PR.

I think you would create a new helper function to know if the fime format is 12-hour-clock or 24-hour-clock. Something like ...

```js
const isTwelveHourFormat = format => (
	endsWith( format, 'A' ) || endsWith( format, 'a' )
);
```

In fact, you could put this function into the `./utils` local lib.

We have to check not only the higher hour value but the minimum value too. For 12-hour-time format, the minimum value is `1`.

> Each period consists of 12 hours numbered: 12 (acting as zero), 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, and 11.

Wiki links
* https://en.wikipedia.org/wiki/12-hour_clock
* https://en.wikipedia.org/wiki/24-hour_clock

iamtakashi commented:
> [... ]Ideally users should only be able to type in 1-12 when in am/pm mode [...]

I agree. 

Is there anything else we should decide?

retrofox commented:
> Is there anything else we should decide?

I don't think @iamtakashi. We will ok with this patch for now I think.

iamtakashi commented:
I've noticed some strange behaviours.

1. I've tried to change the hour from `05` to `06` but PM switched to AM
2. Once the hour hit 12, it doesn't loop back to 01. Although it loops other way around.

![untitled](https://cloud.githubusercontent.com/assets/908665/22894489/0c0896aa-f211-11e6-83a1-12e7114900f5.gif)

iamtakashi commented:
We also need to change the font case for AM/PM (both the top and the bottom) accordingly to the time format setting.

<img width="238" alt="screen shot 2017-02-13 at 17 28 08" src="https://cloud.githubusercontent.com/assets/908665/22894787/2e87bca0-f212-11e6-9bc2-b2c4bc9c4598.png">

obenland commented:
> I've noticed some strange behaviours.
Working on it. Keep those coming, they're helpful

iamtakashi commented:
Not like the min, the hour doesn't seem to take my input with a keyboard.

![untitled](https://cloud.githubusercontent.com/assets/908665/22895071/29adafa4-f213-11e6-9ad9-4f83f0682c83.gif)

obenland commented:
@iamtakashi @retrofox Would you mind testing again with my latest changes?

obenland commented:
> We also need to change the font case for AM/PM (both the top and the bottom) accordingly to the time format setting.

I'd consider that part of a new PR once #10400 is merged and we have a convenient way of converting PHP time formats to Moment formats

retrofox commented:
> I'd consider that part of a new PR once #10400 is merged and we have a convenient way of converting PHP time formats to Moment formats

It doesn't straight depend on #10400 since doesn't matter where the time format edition is done. Regarding the implementation, the primary thing is where the time format is defined, and this is already stored in the state tree.
IMO we should implement state selectors to pick up the needed data.

obenland commented:
> It […] doesn't matter where the time format edition is done.

I'm not sure I understand, could you elaborate? It's my understanding that Moment uses a different kind of format than we have saved in state, [which #10400 provides a mapper for](https://github.com/Automattic/wp-calypso/pull/10400/files#diff-df1fc8d02d6305e47e8598afe79e9df8R339).

> IMO we should implement state selectors to pick up the needed data.

Part of this PR is [a single setting selector](https://github.com/Automattic/wp-calypso/pull/11346/files#diff-46fa1648a32654fbc538a2431c89b2d1) that can be used [to retrieve specific settings](https://github.com/Automattic/wp-calypso/pull/11346/files#diff-084c8c060359137050c43c7503d44863R247). Is there anything I'm missing there?

iamtakashi commented:
Thanks for the update, @obenland!

I can see two of the issues I mentioned have been resolved, but I find something else.... When I increment the hour, the date on the calendar corresponds to it as expected, but when I do the other way around, the date stays the same and the hour is just looping. (I hope the gif helps to illustrate the issue) 

![untitled](https://cloud.githubusercontent.com/assets/908665/22905086/d988a94c-f236-11e6-8833-7b268dd314b6.gif)

obenland commented:
Yes, I'm not entirely sure what to do about that, at this point I suspect it being a Moment.js issue since we only update the time, never the date. Maybe @retrofox knows what that might be caused by?

retrofox commented:
> I'm not sure I understand, could you elaborate?

For being a bit clumsy here. Your code is ok for me.

> It's my understanding that Moment uses a different kind of format than we have saved in state, which #10400 provides a mapper for.

Yup, that's right. But although this `phpToMomentMapping()` function is into #10400 it doesn't strictly belong to the site-settings page. So we could add this helper in your patch.

> Part of this PR is a single setting selector that can be used to retrieve specific settings. Is there anything I'm missing there?

Not at all. It's perfect. What I meant that you could work with the selectors and helper functions to process the data before to bring up to the components.
So you could get easily whatever you need to finish this improvement.

But ... maybe you're right. we can wait to #10400 and iterate later.

iamtakashi commented:
@retrofox Any thoughts on the question from @obenland?

If this isn't possible to fix now, can we compromise it so that both ways don't change the date so that at least it's consistent?

obenland commented:
@iamtakashi Latest commit should fix the date line bug

iamtakashi commented:
> @iamtakashi Latest commit should fix the date line bug
Awesome! I can confirm that.

So, We'll revisit the font casing issue when #10400 lands, correct? Should we ask someone else for a review before merging maybe @rralian?

> Fixes #10762.
Just noting that this PR is a part of fix for the issue, but it shouldn't close it since the time format at the top of the calendar currently stays the same.

retrofox commented on client/components/post-schedule/clock.jsx:
```diff
@@ -191,4 +247,8 @@ PostScheduleClock.defaultProps = {
+	( state, { siteId } ) => ( {
+		timeFormat: getSiteSetting( state, siteId, 'time_format' ),
+	} ),
+)( localize( PostScheduleClock )) ;
```
extra space here

retrofox commented:
I keep seeing a similar erratic behavior here:
![post-schedule-time-counter](https://cloud.githubusercontent.com/assets/77539/22928055/3c97860a-f293-11e6-9691-11d7bbd88ea0.gif)

retrofox commented on client/components/post-schedule/clock.jsx:
```diff
@@ -48,6 +48,11 @@ class PostScheduleClock extends Component {
 
+		// Snap back a day, when looping through date line.
+		if ( 0 === value && 1 === operation ) {
+			modifiers.date = this.props.date.get( 'date' ) - 1;
```
I'm not sure if is ok add `date` key into the modifiers. I think we should just keep working with `hour` and `minute`. Why not just reset the hour value?
Beyond of this, I've found this bug:

`Initial date`: Day 15
`hour`: decrements from `1` to `0`

![image](https://cloud.githubusercontent.com/assets/77539/22928433/2f5b1522-f295-11e6-926e-bc8495670b5e.png)

Day number shouldn't change?

obenland commented:
Thanks for reviewing and testing @retrofox! I fixed the date line bug for both 12- and 24-hour time formats.

The erratic behavior [you mentioned](https://github.com/Automattic/wp-calypso/pull/11346#issuecomment-279688258) is actually built into moment. When you increment `hour` past midnight it will increase `date`as well. What it doesn't do though is to apply the same behavior [when decrementing `hour`](https://github.com/Automattic/wp-calypso/pull/11346#issuecomment-279535881), which I attempted to fix here. After my latest commit it should work  for  both time formats.

iamtakashi commented:
I didn't notice this before but now the min skips some numbers when we increase/decrease it. This seems to occur when the setting with AM/PM. I don't see this behaviour when the setting is 24hr format.

![untitled](https://cloud.githubusercontent.com/assets/908665/22939707/8a87c4d0-f2d7-11e6-890f-df60bc10703d.gif)

obenland commented:
@iamtakashi I didn't restrict the 12-hour magic to just hour inputs, latest commit fixes that

iamtakashi commented:
I'm sorry I didn't catch this earlier, but now the min freezes once I input an erroneous number like `60`.

![untitled](https://cloud.githubusercontent.com/assets/908665/22944195/cf454786-f2e7-11e6-9f8f-839ed2e63bde.gif)

obenland commented:
> I'm sorry I didn't catch this earlier, but now the min freezes once I input an erroneous number like 60.

FWIW, that should be pre-existing and not a regression of this PR.

As you said, It's an erroneous number, it can't really update to it. What would you expect to happen instead?

iamtakashi commented:
The big issue is that it locks out itself and it won't take any input once I put an erroneous number.

I would expect how the hour handles.

iamtakashi commented:
But if the issue is out of scope of this PR, I understand that it should be fixed in a separate issue or PR.

rralian commented:
> But if the issue is out of scope of this PR, I understand that it should be fixed in a separate issue or PR.

@iamtakashi can you create that issue?

iamtakashi commented:
Filed a new issue for the frozen input fields #11409

obenland commented:
@iamtakashi @rralian I had actually worked on that last night, didn't get to commit and push though.

Does it feel better now?

retrofox commented:
I've been testing, and it looks nice. Are we agree that if the hour passes 24 the day will increase as well? 

![post-schedule-time-counter-04](https://cloud.githubusercontent.com/assets/77539/22995201/b44b31a4-f3a8-11e6-8891-86651801c692.gif)

Same question when the hours' number decrease.

iamtakashi commented:
> Does it feel better now?

Definitely. Amazing.

> Are we agree that if the hour passes 24 the day will increase as well?

I'm favour of this behaviour.

obenland commented:
@retrofox What do you think? Should we pull the trigger?

obenland commented:
@iamtakashi Does it still need design review?

iamtakashi commented:
> @iamtakashi Does it still need design review?

@folletto or @drw158 Could you take a look? 
Note: The AM/PM selector only appears when your time setting is one of the formats with AM/PM.

gwwar commented on client/components/post-schedule/utils.js:
```diff
@@ -1,9 +1,32 @@
+	endsWith( timeFormat, 'a' ) ||
+
+	// Uppercase Ante meridiem and Post meridiem.
+	endsWith( timeFormat, 'A' )
```
Optional, but this can probably be shortened to:
```javascript
/[gh]|[aA]$/
```

👍  The comments are still super helpful though

gwwar commented on client/components/post-schedule/clock.jsx:
```diff
@@ -81,6 +108,28 @@ class PostScheduleClock extends Component {
+	 * @param {Number}  hour The hour to convert.
+	 * @return {Number}      The converted hour.
+	 */
+	maybeConvertHour( hour ) {
```
I really dislike the maybe naming pattern. Perhaps something a bit more verbose? `convert12HourTo24HourTime` and move the is12Hour check to the caller.

gwwar commented:
I'd prefer if modifying time doesn't modify date, as we can get into some weird behavior. For example by shifting am/pm values. I can defer though if folks think it's acceptable and we want to get this out first.

![schedule](https://cloud.githubusercontent.com/assets/1270189/23002991/a12d21fa-f3a2-11e6-81c3-9e3aea79a230.gif)

retrofox commented:
> I'd prefer if modifying time doesn't modify date, as we can get into some weird behavior. For example by shifting am/pm values. I can defer though if folks think it's acceptable and we want to get this out first.

I'd prefer to keep the same day beyond the time modification, especially faster read this argument.

folletto commented:
I did a quick review and I found some odd behaviours:

1. The AM/PM advance of a day mentioned above

2. Deleting "12" moves to "01" and then deleting again moves to "12"... is the HTML standard behaviour? Feels a bit odd that I can't clear the form and type, even if I notice that if I ignore what I see and just type it works.
![cap-](https://cloud.githubusercontent.com/assets/4389/23020602/b759ff32-f43f-11e6-8f6e-ab1e93485c10.gif)

3. Doesn't seem to work on a Jetpack site (shows the "am" date above, but below the form doesn't show up)

That said, I'd probably try to push this live, and then review these behaviours separately, as it seems an improvement already as-is.

iamtakashi commented:
Thank you, @folletto!

> Are we agree that if the hour passes 24 the day will increase as well?

The consensus is modifying time doesn't modify date. The example from Kerry was good to exhibit a situation that can be really confusing. So 👍 to keep date alone.

drw158 commented:
I found the same issues as Davide

https://github.com/Automattic/wp-calypso/pull/11346#issuecomment-280317756

Trying to delete the 12 In the hour slot felt especially broken and I think it should be fixed.

Probably unrelated to this PR, but I assume because of daylight savings, this notice at the bottom is misleading:

![screen shot 2017-02-16 at 8 55 43 am](https://cloud.githubusercontent.com/assets/618551/23026271/51b99df4-f426-11e6-8571-5bf5c631a24e.png)

Set the date far in the future or past, and it incorrectly states that my site's timezone is off 1 hour from my local timezone. It should be smart enough to detect DST.

iamtakashi commented:
Thanks, @drw158!

> Probably unrelated to this PR, but I assume because of daylight savings, this notice at the bottom is misleading: Set the date far in the future or past, and it incorrectly states that my site's timezone is off 1 hour from my local timezone. It should be smart enough to detect DST.

Nice catch! I can see this behaviour in production too. Could you create a new issue so that we can fix it separately? Thank you!

obenland commented:
> Deleting "12" moves to "01" and then deleting again moves to "12"... is the HTML standard behaviour? 

@folletto @drw158 It is not standard HTML behavior, on every key press the component is trying to make sense of the input. Deleting the `2` of `12` leaves us with `1`, therefore `01`. Deleting the `1` leaves us with `0`, which in a 12-hour format is `12`.

I took some time to explore what it would take to change this, but it is more involved than just exiting early when we detect the delete button was used. Since this is not a regression caused by this ticket, I'd recommend opening a new issue if you feel this really needs fixing.

gwwar commented:
+1 for moving existing quirks/bugs into separate issues/PRs to address them

obenland commented:
@gwwar @retrofox Just pushed an update to not change dates when crossing date lines. I _should_ have addressed all concerns at this point? Would you mind double checking one last time?

gwwar commented:
Thanks @obenland, behavior looks much better 👍  I tested this on a site with both 12 and 24 hour time. If we can get another set of :eyes: on this to verify, let's 🚢  today, otherwise, let's wait to land until tomorrow.
